src/cryptonote_core/oxen_name_system.cpp

```diff
@@ -830,40 +836,93 @@ bool validate_lns_name(mapping_type type, std::string name, std::string *reason)
+      return false;
+  }
+  else if (type == mapping_type::wallet)
+  {
+    // WALLET
+    // Name has to start with a (alphanumeric or underscore), and can have (alphanumeric, hyphens or underscores) in between and must end with a (alphanumeric or underscore)
+    // ^[a-z0-9_]([a-z0-9-_]*[a-z0-9_])?$
+    // Must start with (alphanumeric or underscore)
+    if (check_condition(!char_is_alphanum_or<'_'>(name_view.front()), reason, "ONS type=", type, ", specifies mapping from name->value where the name does not start with an alphanumeric or underscore character, name=", name))
       return false;
     name_view.remove_prefix(1);
 
     if (!name_view.empty()) {
       // Must NOT end with a hyphen '-'
-      if (check_condition(!char_is_alphanum_or<'_'>(name_view.back()), reason, "LNS type=", type, ", specifies mapping from name->value where the last character is a hyphen '-' which is disallowed, name=", name))
+      if (check_condition(!char_is_alphanum_or<'_'>(name_view.back()), reason, "ONS type=", type, ", specifies mapping from name->value where the last character is a hyphen '-' which is disallowed, name=", name))
         return false;
       name_view.remove_suffix(1);
     }
 
     // Inbetween start and preceding suffix, (alphanumeric, hyphen or underscore) characters permitted
     if (check_condition(!std::all_of(name_view.begin(), name_view.end(), char_is_alphanum_or<'-', '_'>),
-          reason, "LNS type=", type, ", specifies mapping from name->value where the name contains more than the permitted alphanumeric, underscore or hyphen characters, name=", name))
+          reason, "ONS type=", type, ", specifies mapping from name->value where the name contains more than the permitted alphanumeric, underscore or hyphen characters, name=", name))
       return false;
   }
   else
   {
-    MERROR("Wallet names not yet implemented");
+    MERROR("Type not implemented");
     return false;
   }
 
   return true;
 }
 
+std::optional<cryptonote::address_parse_info> encrypted_wallet_value_to_info(std::string name, std::string encrypted_value, std::string nonce)
+{
+
+  std::string lower_name = tools::lowercase_ascii_string(std::move(name));
+  mapping_value record(oxenmq::from_hex(encrypted_value), oxenmq::from_hex(nonce));
+  record.decrypt(lower_name, mapping_type::wallet);
+
+  auto iter = std::next(record.buffer.begin(),1);
+  cryptonote::address_parse_info addr_info{0};
+  addr_info.has_payment_id = false;
+  addr_info.is_subaddress = false;
+  std::memcpy(&addr_info.address.m_spend_public_key.data, &*iter, 32);
+  std::advance(iter,32);
+  std::memcpy(&addr_info.address.m_view_public_key.data, &*iter, 32);
+  if (record.buffer[0] == 0x2) {
+    std::advance(iter,32);
+    std::copy_n(iter,8,addr_info.payment_id.data);
+    addr_info.has_payment_id = true;
+  } else if (record.buffer[0] == 0x1) {
+    addr_info.is_subaddress = true;
+  }
+
```

Comment:
A good chunk of this code looks identical to the earlier code I commented on; can you extract the common bits into a single function to reduce the duplication?